File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.java
// Last updated: 7/2/2025, 11:52:28 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int maxi[] = new int [1];        
        help(root , maxi);
        return maxi[0];
    }

    public int help(TreeNode root, int[] maxi)
    {
        // leaf nodes -> nodes
        if(root == null) return 0;

        int left = help(root.left , maxi);
        int rght = help(root.right , maxi);

        // at each recursive call check if it greater than 
        // maxi , update the global maxi
        maxi[0] = Math.max(maxi[0] , left + rght);

        return 1+Math.max(left , rght);
    }
}

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.cpp
// Last updated: 7/2/2025, 11:52:25 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int d = INT_MIN;
    int diameterOfBinaryTree(TreeNode* root) {
        // might be the chance we can get largest diameter at some node as parent in the tree
      int dd = help(root);
      return d;

    }

    int help(TreeNode *root)
    {
      if(root == NULL) return 0;

      int left = help(root->left);
      int right = help(root->right);

      d = max(d, left+right);

      return 1+max(left, right);
    }
};